### DISCUSSION TITLE
Is there a reason the new attribute for the `<Link>` component is called `legacyBehaviour`?

### DISCUSSION THREAD

**Discussion by nbouvrette** [vercel/next.js#42248] [Help] [answered]
We're trying to upgrade to Next.js 13, and we use a lot of components from Chakra-UI which are children of `<Link>` components.

Unless we misunderstood the only way to make it work, since most of our components use their own style, we would need to add `legacyBehaviour` almost on all our links?

I am a bit worried because it has the name `legacy`. We were also wondering why it's legacy and why can't Next.js support this transparently if we add child elements rather than having to add this property?

**Comment by premier213** [comment]
I think the flow of the work is not suitable for chakraUI, because better use the HOC for all of them, whereas when changing HOC, on another hand your need to change serversideprops and other... new components for this better use of a new solution.

**Reply by nbouvrette** [reply]
Not sure what do you mean by "HOC" ? I'm talking about very basic Chakra components like this here:

```ts
<Link href={href} passHref={true}>
  <Button as="a" {...buttonProps}>
    My Link
  </Button>
</Link>
```

**Comment by premier213** [comment]
use this approach 

![image](https://user-images.githubusercontent.com/4321174/199060732-e597ada5-041a-449a-810d-4118d4bdc80c.png)

**Reply by nbouvrette** [reply]
Ok yeah this is something we could definitely use, but it doesn't answer my other questions:

- why is this called a `legacyBehavior` (even if we wrap it into an HOC I'm worried at some point that the "legacy" will no longer be supported)
- why didn't Next.js decide to make this transparent instead of a breaking change?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The behaviour change has been out since 12.x :), but it was controlled through flags.

It is called legacy because since Next 9.5.x, the `next/link` module could only handle clickable children, or strings. Yes, all along since 9.5, children could've been a string and you wouldn't need the wrapping `anchor` tag, at least not until you needed attributes.

Now you can pass whatever, and they'll wrap it with an anchor tag. 

There's also the problem with classNames, and other attributes that the `next/link` considered out of its own scope.

At the end of the day the `next/link`, exists purely to tell the client side router about SPA transitions.

For your case, you could go to `next.config.js` and tell Next that you wish to keep on going with the legacy behaviour by:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  experimental: { appDir: true, newNextLinkBehavior: false },
};

module.exports = nextConfig;

```

**Reply by nbouvrette** [reply]
Ok, but why isn't the new behavior able to gracefully deal with child `<a>` links without having to refactor a lot of code? I'm just failing to see the reason for this forcing us to use `legacyBehavior` or `newNextLinkBehavior`?

**Reply by icyJoseph** [reply]
I think it is a mix of being careful with such hot-code path as the router link element. 

IIRC there's a bunch of considerations to make because of `cloneElement` usage (in old behavior) and the injection of props to the child, but I can't give you a clear cut answer. 

There's also the lack of `forwardRef` on `next/link` prior to this changes. One had to pass the ref to the child.

Also the new link can take multiple children whereas the old one allowed at maximum one. 

Still, is it possible to technically do a swap, since the API is nearly identical? Perhaps, but they went for an incremental adoption path to prevent edge-cases from breaking applications.

Maybe reading through here you can see a bit more of their reasoning when choosing this way: https://github.com/vercel/next.js/pull/36436#issuecomment-1108663076 too bad the thread is locked 😅

**Reply by nbouvrette** [reply]
Thanks for providing these details.

I made a few tests myself and I really don't get it.. you can add `<div>` or whatever HTML elements inside a `<Link>` which is not even [valid HTML](https://stackoverflow.com/questions/4149876/xhtml-what-elements-are-allowed-within-the-a-element).

In cases where people use frameworks similar to Chakra, we often use pre-built components with pre-built styles/behaviors. Having a child `<a>` link seems to make sense and I still don't understand why Next.js decided to force us to use `legacyBehavior`. The only thing I can think of is to avoid compatibility issues. But even then, technically you could read the props from the children, and if there is a `<a>` either use that object directly or clone its props on your new object.

But still, I would have hoped that they provide a new option like `useLinkWithAnchor` that is a way to say "hey I know what I am doing, just pass all the props to the existing `<a>` instead of creating a new one. This would have avoided

**Comment by nbouvrette** [comment]
By the way, if this helps anyone, there is no need to use `legacyBehavior` - the official solution seems to be to replace `as={'a'}` by `as={Link}` and remove the extra `<Link>` elements.

Ref here: https://github.com/vercel/next.js/pull/36436#issuecomment-1145564252
